### DISCUSSION TITLE
White Screen ESP 32 Dev Kit 1 + ILI9341

### DISCUSSION THREAD

**Discussion by rickyanwar** [Bodmer/TFT_eSPI#950] [General]
Nice job,
I am using this library with esp32 + ILI9341 2.4 inch i was configure User_Setup like this, using the example but display still white screen

Link Lcd
https://shopee.co.id/product/129766956/2564805341?smtt=0.1066989-1611331272.9
```
#define ILI9341_DRIVER

#define TFT_MISO 19
#define TFT_MOSI 23
#define TFT_SCLK 18
#define TFT_CS 15 // Chip select control pin//
#define TFT_DC 2 // Data Command control
#define TFT_RST 4 
```

**Comment by Bodmer** [comment]
This is a suitable setup based on your pin selection:
```
#define ILI9341_DRIVER

#define TFT_MOSI 23
#define TFT_MISO 19
#define TFT_SCLK 18
#define TFT_CS   5  // Chip select control pin
#define TFT_DC   2  // Data Command control pin
#define TFT_RST  4  // Reset pin (could connect to RST pin)

// Optional - define for XPT2046 touch controller
//#define TOUCH_CS 22     // Chip select pin (T_CS) of touch screen

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

// Comment out the #define below to stop the SPIFFS filing system and smooth font code being loaded
// this will save ~20kbytes of FLASH
#define SMOOTH_FONT

#define SPI_FREQUENCY  27000000 // Actually sets it to 26.67MHz = 80/3
// #define SPI_FREQUENCY  40000000 // Maximum recommended for ILI9341

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  10000000

// The XPT2046 requires a lower SPI clock rate of 2.5MHz so we define that here:
#define SPI_TOUCH_FREQUENCY  400000
```

Exactly how have you connected it up, list all connections (which pin to which pin)?
Arre the connections good (push-in wire breadboard connections can be unreliable, keep connections short)?
 Are you using the Arduino IDE - which version?
Which version of the TFT_eSPI library?

**Reply by rickyanwar** [reply]
Thank you for answer my question
I trying setup like your comment but it still white screen, 
yes it pins to pin i following this tutorial 
https://www.youtube.com/watch?v=rq5yPJbX_uk&t=214s

using `Library version 2.3.59`, i was trying downgrade version but still same
`Arduino IDE Version 1.8.13` using example `TFT_Meters`

![photo_2021-01-23_23-26-01](https://user-images.githubusercontent.com/36995966/105606139-7d33e380-5dd2-11eb-85be-79b0ea7dcb8c.jpg)
Sorry cant screen shoot while dropdown menu in Linux, so I capture it with my phone

**Comment by Bodmer** [comment]
The tutorial looks OK.
I suggest you now run the Test and diagnostics sketch : Read_User_Setup

This will output the settings picked up by the compiler to the Serial Monitor window, and you can check these are correct. Post the output here if you are not sure.

**Comment by rickyanwar** [comment]
this is log Read_User_Setup, I think it same with my setting
```
[code]
TFT_eSPI ver = 2.3.59
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9341
Display width  = 240
Display height = 320

MOSI    = GPIO 23
MISO    = GPIO 19
SCK     = GPIO 18
TFT_CS   = GPIO 15
TFT_DC   = GPIO 2
TFT_RST  = GPIO 4

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 27.00
[/code]

```

**Comment by Bodmer** [comment]
That looks good, you now need to check the wiring. I have connected up my board to the display and used the same settings and it works fine.  here is the diagnostic output:
```
[code]
TFT_eSPI ver = 2.3.59
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9341
Display width  = 240
Display height = 320

MOSI    = GPIO 23
MISO    = GPIO 19
SCK     = GPIO 18
TFT_CS   = GPIO 15
TFT_DC   = GPIO 2
TFT_RST  = GPIO 4

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 27.00
[/code]
```
Here is the minimal wiring setup that I suggest you follow. This works fine:
![ESP32_1](https://user-images.githubusercontent.com/15824805/105615992-72298500-5dcb-11eb-862e-5e40df693f62.jpg)

![ESP32_2](https://user-images.githubusercontent.com/15824805/105615993-781f6600-5dcb-11eb-999d-081b9a44382c.jpg)

Unfortunately I cannot replicate your problem so will not be able to help further.

**Comment by Bodmer** [comment]
This is a copy of my setup file:
```
#define ILI9341_DRIVER

#define TFT_MOSI 23
#define TFT_MISO 19
#define TFT_SCLK 18
#define TFT_CS   15 // Chip select control pin
#define TFT_DC   2  // Data Command control pin
#define TFT_RST  4  // Reset pin (could connect to RST pin)

// Optional - define for XPT2046 touch controller
//#define TOUCH_CS 22     // Chip select pin (T_CS) of touch screen

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

// Comment out the #define below to stop the SPIFFS filing system and smooth font code being loaded
// this will save ~20kbytes of FLASH
#define SMOOTH_FONT

#define SPI_FREQUENCY  27000000 // Actually sets it to 26.67MHz = 80/3
// #define SPI_FREQUENCY  40000000 // Maximum recommended for ILI9341

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  10000000

// The XPT2046 requires a lower SPI clock rate of 2.5MHz so we define that here:
#define SPI_TOUCH_FREQUENCY  400000
```

**Comment by rickyanwar** [comment]
This is my setup
```
[code]
TFT_eSPI ver = 2.3.59
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9341
Display width  = 240
Display height = 320

MOSI    = GPIO 23
MISO    = GPIO 19
SCK     = GPIO 18
TFT_CS   = GPIO 15
TFT_DC   = GPIO 2
TFT_RST  = GPIO 4

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 27.00
[/code]
```

This is my wiring but still same blank

![IMG_20210124_153457](https://user-images.githubusercontent.com/36995966/105624144-5faa5c80-5e5a-11eb-9f36-5711ca6f3574.jpg)
![IMG_20210124_153611](https://user-images.githubusercontent.com/36995966/105624146-63d67a00-5e5a-11eb-8c76-0a1c4661b3cb.jpg)
![IMG_20210124_153450](https://user-images.githubusercontent.com/36995966/105624148-65a03d80-5e5a-11eb-947a-2028f23b62b6.jpg)

**Comment by Bodmer** [comment]
Remove the esp32 module from the baseboard, wire TFT direct to pins of esp32 module. Power the esp32 module from USB only.

I suspect the baseboard is affecting operation, so provide a web link and schematic of the KelasRobot base board.

**Comment by Bodmer** [comment]
Are you sure the 3V3 pins on the baseboard are supplying power?

**Comment by rickyanwar** [comment]
![IMG_20210124_231003](https://user-images.githubusercontent.com/36995966/105634534-6a381680-5e99-11eb-8812-9b5f93676b66.jpg)
Yes this 3 V pin

**Comment by Bodmer** [comment]
It may not be sourcing power unless the board is supplied via 7 to 12V via the DC round plug...

**Comment by rickyanwar** [comment]
sorry i am new microcontroller. thanks a lot for answering my question, the pin in VCC and led it must in single pin 3v ? or can be different 1 for VCC pin and 1 for LED pin? because I see in a tutorial it 1 pin for VCC and led pin

**Comment by Bodmer** [comment]
It can be different.  I think maybe the base board supply is NOT active unless it has the DC 7V-12V connected. That is why I asked for a web link to the vendors base board page.

Remove the ESP32 and wire direct to the ESP32 like I have. It is always best to start with a minimal complexity system.

**Comment by rickyanwar** [comment]
i was trying using adapter 12v 2a but is not work,
trying not using baseboard same is not work
hmm whats wrong with my TFT :(

If my lcd is broken why it just blank? You can suggest how to detect if it broken?

**Comment by Bodmer** [comment]
You can try to see if the TFT registers can be read, use this sketch and look at the Serial Monitor output:
```

//ESP32
#define TFT_MOSI  23
#define TFT_MISO  19
#define TFT_SCK 18
#define TFT_CS 15
#define TFT_DC  2
#define TFT_RESET 4


uint32_t readwrite8(uint8_t cmd, uint8_t bits, uint8_t dummy)
{
    uint32_t ret = 0;
    uint8_t val = cmd;
    int cnt = 8;
    digitalWrite(TFT_CS, LOW);
    digitalWrite(TFT_DC, LOW);
    for (int i = 0; i < 8; i++) {   //send command
        digitalWrite(TFT_MOSI, (val & 0x80) != 0);
        digitalWrite(TFT_SCK, HIGH);
        digitalWrite(TFT_SCK, LOW);
        val <<= 1;
    }
    if (bits == 0) {
        digitalWrite(TFT_CS, HIGH);
        return 0;
    }
    digitalWrite(TFT_DC, HIGH);
    for (int i = 0; i < dummy; i++) {  //any dummy clocks
        digitalWrite(TFT_SCK, HIGH);
        digitalWrite(TFT_SCK, LOW);
    }
    for (int i = 0; i < bits; i++) {  // read results
        ret <<= 1;
        if (digitalRead(TFT_MISO)) ret |= 1;;
        digitalWrite(TFT_SCK, HIGH);
        digitalWrite(TFT_SCK, LOW);
    }
    digitalWrite(TFT_CS, HIGH);
    return ret;
}

void showreg(uint8_t reg, uint8_t bits, uint8_t dummy)
{
    uint32_t val;
    val = readwrite8(reg, bits, dummy);

    Serial.print("Register 0x");
    if (reg < 0x10) Serial.print("0");
    Serial.print(reg , HEX);
    Serial.print(": 0x");
    if (val < 0x10) Serial.print("0");
    Serial.println(val, HEX);
}

void setup() {
    Serial.begin(115200);

    delay(100);
    
    Serial.println("\n\nTFT driver register values:");
    Serial.println("===========================");
    digitalWrite(TFT_CS, HIGH);
    //    digitalWrite(TFT_SCK, HIGH);
    pinMode(TFT_CS, OUTPUT);
    pinMode(TFT_SCK, OUTPUT);
    pinMode(TFT_MOSI, OUTPUT);
    pinMode(TFT_MISO, INPUT_PULLUP);
    pinMode(TFT_DC, OUTPUT);
    pinMode(TFT_RESET, OUTPUT);
    digitalWrite(TFT_RESET, HIGH);
    digitalWrite(TFT_RESET, LOW);   //Hardware Reset
    delay(50);
    digitalWrite(TFT_RESET, HIGH);
    showreg(0x01, 0, 0);            //Software Reset
    delay(100);

    showreg(0x04, 24, 1);   //RDDID
    showreg(0x09, 32, 1);   //RDSTATUS
    showreg(0x0A, 8, 0);
    showreg(0x0B, 8, 0);
    showreg(0x0C, 8, 0);
    showreg(0x0D, 8, 0);
    showreg(0x0E, 8, 0);
    showreg(0x0F, 8, 0);
    showreg(0x2E, 24, 8);   //readGRAM
    showreg(0x3F, 32, 1);   //readGRAM
    showreg(0xDA, 8, 0);
    showreg(0xDB, 8, 0);
    showreg(0xDC, 8, 0);

    Serial.println("===========================");
}

void loop() {
}
```

This is my output:
```
TFT driver register values:
===========================
Register 0x01: 0x00
Register 0x04: 0x00
Register 0x09: 0x610000
Register 0x0A: 0x08
Register 0x0B: 0x00
Register 0x0C: 0x06
Register 0x0D: 0x00
Register 0x0E: 0x00
Register 0x0F: 0x00
Register 0x2E: 0x545454
Register 0x3F: 0xFFFFFFFF
Register 0xDA: 0x00
Register 0xDB: 0x00
Register 0xDC: 0x00
===========================

```

**Comment by rickyanwar** [comment]
this is my log, what you think?
```

TFT driver register values:
===========================
Register 0x01: 0x00
Register 0x04: 0x00
Register 0x09: 0x00
Register 0x0A: 0x00
Register 0x0B: 0x00
Register 0x0C: 0x00
Register 0x0D: 0x00
Register 0x0E: 0x00
Register 0x0F: 0x00
Register 0x2E: 0x00
Register 0x3F: 0x80000000
Register 0xDA: 0x00
Register 0xDB: 0x00
Register 0xDC: 0x00
===========================

```

**Comment by Bodmer** [comment]
Those results indicate there is no communication happening with the display chip for some reason. Since the TFT_eSPI library is not used for the test this indicates that the library will not be able to talk to the chip either.

This might not work but try it: The displays are fitted with a regulator (U1 on PCB) which is designed to accept 5V input and output 3.3V to the display. These are normally "Low Drop Out" (LDO) which means that if you connect 3.3V to them they still output about 3.0V and this is sufficient for the display chip. Your board may be fitted with a regulator that has a higher voltage loss, so it is worth trying to power the display from 5V connected to the VCC pin.  The other option is to bypass the regulator U1 by soldering a link (or adding a big blob of solder) to short across J1 pads. This then links VCC direct to the TFT chip and thus VCC must only be 3.3V (5V with that link will damage the chip).

I have many displays and rarely have a faulty one. The faulty one had a cracked screen.

It could be that the TFt has a competely different controller fitted but I would expect to see some values being returned by the simple test sketch.

It may be the display chip has been blown by accidental connections during testing and debugging, but I have found they are quite robust to moderate abuse.

Lastly it could be that your wires are faulty, you can get bad crimps in those link wire connectors and so there is not connection even though it is physically fitted to the pins.

**Comment by rickyanwar** [comment]
Sorry to trouble you, 
I never trying with 5v to vcc
And i was trying to change my wires with antoher, i was change 5 times

this is file given by the seller hope you can help me to solve this problem

```

This is a new 2.4 inch serial SPI color display module without touch function.
Support analog SPI and hardware SPI.
Designed with one socket for sd card.
With PCB plate (including power supply IC, SD), compatible with 5110 interface.
It can be driven with at least four IO.

Specifications:
Resolution: 240x320
Driver IC: ILI9341
Input Voltage: 5V/3.3V
Quantity: 1
Document Download:
http://tftdata.tjc1688.com/2.4_spi/2.4_spi.rar
Package Includes:
```

**Reply by rickyanwar** [reply]
TFT      =>  ESP32
CS        =>  D15
Reset     => D4
DC         => D2
SDI (mosi)      =>  D23
SCK      => D18
LED      => 3V
SDO (miso)     =>19   

i was trying again i got diffrent value,put adapter and usb, i got same log with you hmmm
```


TFT driver register values:
===========================
Register 0x01: 0x00
Register 0x04: 0x00
Register 0x09: 0x610000
Register 0x0A: 0x08
Register 0x0B: 0x00
Register 0x0C: 0x06
Register 0x0D: 0x00
Register 0x0E: 0x00
Register 0x0F: 0x00
Register 0x2E: 0x545454
Register 0x3F: 0xFFFFFFFF
Register 0xDA: 0x00
Register 0xDB: 0x00
Register 0xDC: 0x00
===========================
```

**Reply by Bodmer** [reply]
That is good, the library should work now.

**Reply by rickyanwar** [reply]
Same not work, it just white screen, i want send you video but github cant

**Reply by Bodmer** [reply]
I am not sure exactly what you mean by "i was trying again i got diffrent value,put adapter and usb, i got same log with you" but since you have altered the setup to work with the simple test sketch you have managed to get the required communication.

I don't think a video of a system that does not work will help.  Your main problem seems to be using that baseboard. Try again "without" the base board and get a simple direct display to ESP32 working with that simple sketch. Put that baseboard away in a drawer and do NOT use it for now.

Post pictures of this new simplified ESP32 setup.

**Reply by Bodmer** [reply]
That is the touch controller chip, it is connected to the T-xx pins. Your display does not have a touch screen so the chip is not fitted, that is not the cause of the problem.

**Reply by rickyanwar** [reply]
trying not using baseboard, but a same white screen
wiring 
VCC -> VIN
LCD -> 3V
CS => D15
Reset => D4
DC => D2
SDI (mosi) => D23
SCK => D18
SDO (miso) =>19
and my log 

```
TFT driver register values:
===========================
Register 0x01: 0x00
Register 0x04: 0xF8F8F8
Register 0x09: 0x610000
Register 0x0A: 0x08
Register 0x0B: 0x00
Register 0x0C: 0x06
Register 0x0D: 0x00
Register 0x0E: 0x00
Register 0x0F: 0x00
Register 0x2E: 0x545454
Register 0x3F: 0xFFFFFFFF
Register 0xDA: 0xF8
Register 0xDB: 0xF8
Register 0xDC: 0xF8
===========================
```


![IMG_20210127_011218](https://user-images.githubusercontent.com/36995966/105879167-cedfa680-603c-11eb-8ec5-09c3c380451a.jpg)
![IMG_20210127_011314](https://user-images.githubusercontent.com/36995966/105879205-d901a500-603c-11eb-8951-1d15e43fd728.jpg)
![IMG_20210127_011525](https://user-images.githubusercontent.com/36995966/105879484-3138a700-603d-11eb-8799-bd763daf2d40.jpg)

**Reply by Bodmer** [reply]
The register test sketch will NOT display anything, it is just a SPI bus communication test.

You need to run a TFT_eSPI example now, make sure it is configured correctly.

The good news is that communication with the display looks good.

**Reply by rickyanwar** [reply]
i was trying the example, but nothing shows (White Screen), I always get different value from the register log

```

TFT driver register values:
===========================
Register 0x01: 0x00
Register 0x04: 0x808080
Register 0x09: 0x610000
Register 0x0A: 0x08
Register 0x0B: 0x00
Register 0x0C: 0x06
Register 0x0D: 0x00
Register 0x0E: 0x00
Register 0x0F: 0x00
Register 0x2E: 0x545454
Register 0x3F: 0xFFFFFFFF
Register 0xDA: 0x80
Register 0xDB: 0x80
Register 0xDC: 0x80
===========================


```

**Reply by Bodmer** [reply]
A video of a non-functioning system will not help.

I suspect originally you could not read the registers due to a power supply problem. Now that is fixed and you can read the registers then TFT_eSPI should work fine.

Run the Read_User_Setup again to check the settings picked up by the compiler are the same as the connections to the TFT.  Reduce the SPI clock rate to 10MHz:
#define SPI_FREQUENCY  10000000

This will help if the length of the wires is too long.

I would to delete, reload and re-configure a new version of TFT_eSPI in cse it has been corrupted. Also check you have the latest ESP32 board package.

We have tried hard together to get this working but we are close approaching the point where I would try a new display and new ESP32 board.

I am unable to reproduce the problem you are experiencing with an electrically identical hardware setup.

**Reply by rickyanwar** [reply]
![IMG_20210206_131711](https://user-images.githubusercontent.com/36995966/107113393-a99f3380-6899-11eb-9f22-e8302dd8aa68.jpg)
I was buying new LCD it work

**Comment by rickyanwar** [comment]
![IMG_20210206_131711](https://user-images.githubusercontent.com/36995966/107109759-d9404280-687d-11eb-933c-6124103efab3.jpg)
I buy new LCD it work,

**Comment by Bodmer** [comment]
Yay, ay last!

A duff screen is quite unusual. You were unlucky.

**Reply by rickyanwar** [reply]
Thanks you for help me  😁
